colossalai/context/moe_context.py

```diff
@@ -0,0 +1,139 @@
+import torch
+import torch.distributed as dist
+from .parallel_mode import ParallelMode
+
+
+class MoeInfo:
+    """Moe parallelism information, storing parallel sizes and groups.
+    """
+
+    def __init__(self, ep_size: int, dp_size: int):
+        self.ep_size = ep_size
+        self.dp_size = dp_size
+        self.ep_group = None
+        self.dp_group = None
+
+        # Here we assume tensor parallel size = 1
+        # Otherwise, MoE can't be used
+        # Since TENSOR parallel group and DATA parallel group
+        # have been created, we can use them directly.
+        if ep_size == 1:
+            from colossalai.core import global_context as gpc
+            self.ep_group = gpc.get_group(ParallelMode.TENSOR)
+            self.dp_group = gpc.get_group(ParallelMode.DATA)
+            return
+
+        if dp_size == 1:
+            from colossalai.core import global_context as gpc
+            self.ep_group = gpc.get_group(ParallelMode.DATA)
+            self.dp_group = gpc.get_group(ParallelMode.TENSOR)
+            return
+
+        rank = dist.get_rank()
+        # Create expert parallel group
+        for i in range(dp_size):
+            ranks = [i * ep_size + j for j in range(ep_size)]
+            group = dist.new_group(ranks)
+            if rank in ranks:
+                self.ep_group = group
+
+        # Create data parallel group
+        for j in range(ep_size):
+            ranks = [i * ep_size + j for i in range(dp_size)]
+            group = dist.new_group(ranks)
+            if rank in ranks:
+                self.dp_group = group
+
+
+class MoeContext:
+    """MoE parallel context manager. This class manages different
+    parallel groups in MoE context and MoE loss in training.
+    """
+    __instance = None
+
+    @staticmethod
+    def get_instance():
+        if MoeContext.__instance is None:
+            MoeContext.__instance = MoeContext()
+        return MoeContext.__instance
+
+    def __init__(self):
+        self.world_size = 1
+        # Users may want to set maximum expert parallel size smaller than the world size
+        # since very low bandwidth across nodes may constrain the performance of MoE
+        # When we have a maximum expert parallel size, we have a minimum data parallel size naturally

```

Comment:
what do you want to say? any suggestions to the max_ep_size setting?

Reply:
Just tell the function of max_ep_size